### ISSUE TITLE
Use props `dict` on components, rather than kwargs.

### ISSUE THREAD

**Issue by Archmonger** [reactive-python/reactpy#993]
### Current Situation

There's been a lot of growing pains in the fact that components have a significantly different interface than `reactpy.html` elements.

Developing certain parts of `reactpy` core and/or `reactpy` related utilities has been significantly more difficult due to this interface mismatch.

### Proposed Actions

The new component interface should match the [VDOM interface](https://github.com/reactive-python/reactpy/blob/7428340b44c8db5728fcff3b36908735ecfd4ca3/src/py/reactpy/reactpy/core/types.py#L198-L213), and will look like this:

```python
# `key=...` should be declared within props, identical to the `reactpy.html` interface.
my_component({"prop_1": True, "key": 123}, child_1, child_2)
```

As a part of this change, we should rename `VdomAttributes` to `Props`.

### Do we retain args/kwargs?

We can handle `@component` function definitions a few ways. Some of these ways feel more similar to ReactJS, but have problems with Python type hints. 

1) Force all users to only use kwargs
   - Retain type hints via [PEP692 ](https://peps.python.org/pep-0692/)
2) Allow args/kwargs in function defs. Let Python unpacking operator handle the rest.
   - This would break type hints, since args can't be defined within a TypedDict
3) Force all users to define a `props: dict` as their first parameter
   - However, forcing users to manually define a TypedDict for every component would be extremely inconvenient.

I'm personally leaning towards option 1.

**Comment by rmorshea** [comment]
More precisely the interface should conform to [this protocol](https://github.com/reactive-python/reactpy/blob/main/src/reactpy/core/types.py#L198-L213).

**Comment by Archmonger** [comment]
We can potentially retain type hints using this PEP https://peps.python.org/pep-0692/

**Comment by Archmonger** [comment]
I spent more thought on this issue while rewriting the docs.

@rmorshea We need to decide what we will do with the user's component function defs. See my updates above.

**Comment by rmorshea** [comment]
Note: we'll have to figure out how this will play work if the component function is a method.

**Comment by Archmonger** [comment]
I did some prototyping, and learned it's currently not possible to use `ParamSpec` to provide type hints to a dictionary.

Given that components are functions, it feels natural for users to call them via the standard "python function interface" (AKA arguments within parenthesis)

Additionally, changing the `reactpy.html` interface to kwargs makes the HTML "read wrong", where all HTML properties are now declared at the bottom of the elements, making it almost unreadable to figure out what prop belongs to which element in deeply nested structures.

Unless the Python type hinting space makes some major changes, I think we're forced to continue down the road we've already paved. I've already [raised the issue](https://discuss.python.org/t/type-hints-for-kwarg-overrides/47196) that the Python type hinting system has no way to "mutate" kwargs, so hopefully that gains some interest/traction.

**Comment by rmorshea** [comment]
A [draft of the template strings PEP](https://peps.python.org/pep-0750/) is currently in review so optimizing the interface may not be necessary anyways if that lands.

**Comment by Archmonger** [comment]
Agreed, I recently made that same conclusion while talking about the `reactpy.html` discussion on Discord.
